File: exercise09-A/src/program.cpp
// Exercise 09-A: Traversal

#include <iostream>
#include <queue>
#include <stack>
#include <string>
#include <vector>

// Node structure

template <typename T>
struct Node {
    T value;
    Node *left;
    Node *right;
};

// Pre-defined Trees

Node<char> *AlgorithmTree() {
    return new Node<char>{'A',
                          new Node<char>{'L',
                                         new Node<char>{'O',
                                                        new Node<char>{'H', nullptr, nullptr},
                                                        new Node<char>{'M', nullptr, nullptr}},
                                         new Node<char>{'R', nullptr, nullptr}},
                          new Node<char>{'G',
                                         new Node<char>{'I', nullptr, nullptr},
                                         new Node<char>{'T', nullptr, nullptr}}};
}

// Traversal: BFS (Iterative)

template <typename T>
void bfs(Node<T> *root) {
}

// Traversal: DFS (Iterative)

template <typename T>
void dfs(Node<T> *root) {
}

// Traversal: DFS (Recursive)

template <typename T>
void dfs_recursive(Node<T> *root) {
}

// Main execution

int main(int argc, char *argv[]) {
    auto tree = AlgorithmTree();

    std::cout << "BFS:";
    bfs(tree);
    std::cout << std::endl;
    std::cout << "DFS:";
    dfs(tree);
    std::cout << std::endl;
    std::cout << "DFS:";
    dfs_recursive(tree);
    std::cout << std::endl;

    return EXIT_SUCCESS;
}

File: exercise09-A/src/program.py
#!/usr/bin/env python3

# Exercise 09-A: Traversal

from dataclasses import dataclass
from collections import deque

# Node structure


@dataclass
class Node:
    value: str
    left: "Node" = None
    right: "Node" = None


# Pre-defined Trees

AlgorithmTree = Node(
    "A",
    Node("L", Node("O", Node("H"), Node("M")), Node("R")),
    Node("G", Node("I"), Node("T")),
)

# Traversal: BFS (Iterative)


def bfs(root: Node) -> None:
    queue = deque([root])
    while queue:
        if node := queue.pop():
            print(f" {node.value}", end="")
            queue.appendleft(node.left)
            queue.appendleft(node.right)


# Traversal: DFS (Iterative)


def dfs(root: Node) -> None:
    stack = [root]
    while stack:
        if node := stack.pop():
            print(f" {node.value}", end="")
            stack.append(node.right)
            stack.append(node.left)


# Traversal: DFS (Recursive)


def dfs_recursive(root: Node) -> None:
    if root:
        print(" " + root.value, end="")
        dfs_recursive(root.left)
        dfs_recursive(root.right)


# Main Execution


def main():
    print(f"BFS:", end="")
    bfs(AlgorithmTree)
    print()
    print(f"DFS:", end="")
    dfs(AlgorithmTree)
    print()
    print(f"DFS:", end="")
    dfs_recursive(AlgorithmTree)
    print()


if __name__ == "__main__":
    main()
